source/dub/recipe/packagerecipe.d

```diff
@@ -171,50 +213,105 @@ struct ConfigurationInfo {
+private struct DependencyObject
+{
+    bool optional;
+    @Optional @Name("version")
+    string version_;
+    @Optional string path;
+}
+
+/// Wrapper around `DependencyObject` to do deserialization but still provide
+/// a `Dependency` object to client code.
+private struct YAMLDep
+{
+    private DependencyObject representation;
+    public Dependency actual;
+
+    alias actual this;
+
+    public this (inout(Dependency) dep) inout @safe pure nothrow @nogc
+    {
+        this.actual = dep;
+    }
+
+    public ref YAMLDep opAssign (Dependency dep) return @safe pure nothrow @nogc
+    {
+        this.actual = dep;
+        return this;
+    }
+
+    static YAMLDep fromYAML (scope ConfigParser!YAMLDep p)
+    {
+        import dyaml.node;
+
+        if (p.node.nodeID == NodeID.scalar)
+            return YAMLDep(Dependency(p.node.as!string));
+        auto d = p.parseField!"representation";
+        if (d.path.length)
+        {
+            auto r = YAMLDep(Dependency(NativePath(d.path)));
+            r.optional = d.optional;
+            return r;
+        }
+        else
+        {
+            assert(d.version_.length);
+            auto r = YAMLDep(Dependency(d.version_));
+            r.optional = d.optional;
+            return r;
+        }
+    }
+}
+
 /// This keeps general information about how to build a package.
 /// It contains functions to create a specific BuildSetting, targeted at
 /// a certain BuildPlatform.
 struct BuildSettingsTemplate {
-	Dependency[string] dependencies;
-	BuildSettingsTemplate[string] dependencyBuildSettings;
-	string systemDependencies;
-	TargetType targetType = TargetType.autodetect;
-	string targetPath;
-	string targetName;
-	string workingDirectory;
-	string mainSourceFile;
-	string[string] subConfigurations;
-	string[][string] dflags;
-	string[][string] lflags;
-	string[][string] libs;
-	string[][string] sourceFiles;
-	string[][string] sourcePaths;
-	string[][string] excludedSourceFiles;
-	string[][string] injectSourceFiles;
-	string[][string] copyFiles;
-	string[][string] extraDependencyFiles;
-	string[][string] versions;
-	string[][string] debugVersions;
-	string[][string] versionFilters;
-	string[][string] debugVersionFilters;
-	string[][string] importPaths;
-	string[][string] stringImportPaths;
-	string[][string] preGenerateCommands;
-	string[][string] postGenerateCommands;
-	string[][string] preBuildCommands;
-	string[][string] postBuildCommands;
-	string[][string] preRunCommands;
-	string[][string] postRunCommands;
-	string[string][string] environments;
-	string[string][string] buildEnvironments;
-	string[string][string] runEnvironments;
-	string[string][string] preGenerateEnvironments;
-	string[string][string] postGenerateEnvironments;
-	string[string][string] preBuildEnvironments;
-	string[string][string] postBuildEnvironments;
-	string[string][string] preRunEnvironments;
-	string[string][string] postRunEnvironments;
+	@Optional YAMLDep[string] dependencies;
```

Comment:
do we really need to expose format details to the outside? I think nothing in packagerecipe should expose YAML related stuff

Reply:
I tried very hard to remove any access / need for direct YAML manipulation from Configy.
The idea being you could support more format later. However, there is currently no way to support an "union", as `dependencies` does. Dependencies can be a plain scalar (in which case it's a range specification) or an object. That's where the need for `fromYAML` comes from. I tried adding support for `SumType` in configy but couldn't come up with something that was robust, easy to use, and in the spirit of the library.

Reply:
SumType would probably be best integrated into the language, as it's also part of stdlib. I faced the same issue with JSON-RPC serialization in serve-d - I settled with mir-ion and mir-algebraic for now, added a custom struct for determining the correct struct(s) from various possible structs and am planning to try to get it merged upstream eventually.